### PR TITLE
NodeID and NodePath

### PR DESCRIPTION
**Pull request by bekand**
# Terms

> THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE TERMS OF THE CCLA DATED 2017-11-07 WITH FINOS/LINUX FOUNDATION (FORMERLY THE SYMPHONY SOFTWARE FOUNDATION CCLA).
>
> THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

### CODE REVIEW

File: src/Morphir/IR/NodeId.elm [hunk 1]
```diff
@@ -1,48 +1,754 @@
...
+returnInvalidPathError pathSoFar =
+    case nodePathToString pathSoFar of
+        "" ->
+            Err <| InvalidPath "Path is invalid"
+
+        _ ->
+            Err <| InvalidPath ("Path is invalid after " ++ nodePathToString pathSoFar)
+
+
+getFromList : List NodePathStep -> List NodePathStep -> (NodePath -> NodePath -> a -> Result Error attr) -> List a -> Result Error attr
```
Comment (osborneMs) [inline src/Morphir/IR/NodeId.elm:180]: Add docs on what the various function do. It helps with understanding

File: src/Morphir/IR/NodeId.elm [hunk 2]
```diff
@@ -1,48 +1,754 @@
...
+            list
+                |> List.Extra.getAt n
+                |> Maybe.map (recursiveFunction xs (List.append [ current ] pathSoFar))
+                |> Maybe.withDefault (returnInvalidPathError pathSoFar)
+
+        _ ->
+            returnInvalidPathError pathSoFar
+
+
+getAttribute : Package.Definition attr attr -> NodeID -> Result Error attr
```
Comment (osborneMs) [inline src/Morphir/IR/NodeId.elm:193]: Could you please alias the type `attr`. Its difficult understanding what that is.

File: src/Morphir/CustomAttribute/CustomAttribute.elm
Comment (osborneMs) [inline src/Morphir/CustomAttribute/CustomAttribute.elm:1]: Not sure if this file should still exist or be renamed. Can you please confirm ?

**Review by osborneMs — CHANGES_REQUESTED** [review]

File: src/Morphir/IR/NodeId.elm [hunk 3]
```diff
@@ -1,48 +1,754 @@
...
+
+            else
+                mapToTypeOrValue packageName moduleName localName ""
+
         _ ->
-            Err <| "Invalid NodeId" ++ str
+            returnError
 
 
 nodeIdToString : NodeID -> String
```
Comment (AttilaMihaly) [inline src/Morphir/IR/NodeId.elm:97]: Every exposed function should have some docs.

File: src/Morphir/IR/NodeId.elm [hunk 4]
```diff
@@ -1,48 +1,754 @@
...
 nodeIdToString nodeId =
+    let
+        mapToTypeOrValue packageName moduleName localName suffix nodePath =
```
Comment (AttilaMihaly) [inline src/Morphir/IR/NodeId.elm:100]: Always specify the types. Especially if there's this many arguments.

**Review by AttilaMihaly — CHANGES_REQUESTED** [review]

File: src/Morphir/IR/NodeId.elm [hunk 5]
```diff
@@ -1,48 +1,795 @@
...
+                        (mapAttribute a)
+                        (mapValueAttributesWithNodePathRec mf (pathToMe ++ [ ChildByIndex 0 ]) rec)
+                        (mapDict (pathToMe ++ [ ChildByIndex 1 ]) updateDict)
+    in
+    mapValueAttributesWithNodePathRec mapFunc [] value
+
+
+{-| Map pattern attribute
+-}
+mapPatternAttributesWithNodePath : (NodePath -> attr -> attr2) -> Value.Pattern attr -> Value.Pattern attr2
```
Comment (AttilaMihaly) [inline src/Morphir/IR/NodeId.elm:464]: It feels like this deserves more explanation.

File: src/Morphir/IR/NodeId.elm [hunk 6]
```diff
@@ -1,48 +1,795 @@
...
+mapPatternAttributesWithNodePath mapFunc pattern =
+    mapPatternAttributesWithNodePathRec mapFunc [] pattern
+
+
+mapPatternAttributesWithNodePathRec : (NodePath -> attr -> attr2) -> NodePath -> Value.Pattern attr -> Value.Pattern attr2
```
Comment (AttilaMihaly) [inline src/Morphir/IR/NodeId.elm:469]: Docs are missing

File: src/Morphir/IR/NodeId.elm [hunk 7]
```diff
@@ -1,48 +1,795 @@
...
+{-| Map type attribute
+-}
+mapTypeAttributeWithNodePath : (NodePath -> attr -> attr2) -> Type attr -> Type attr2
+mapTypeAttributeWithNodePath mapFunc tpe =
+    mapTypeAttributeWithNodePathRec mapFunc [] tpe
+
+
+{-| Map value attribute
+-}
+mapValueAttributesWithNodePath : (NodePath -> attr -> attr2) -> Value attr attr -> Value attr2 attr2
```
Comment (AttilaMihaly) [inline src/Morphir/IR/NodeId.elm:325]: Needs more explanation.

File: src/Morphir/IR/NodeId.elm [hunk 8]
```diff
@@ -1,48 +1,795 @@
...
+                (mapTypeAttributeWithNodePathRec mf (pathToMe ++ [ ChildByIndex 0 ]) input)
+                (mapTypeAttributeWithNodePathRec mf (pathToMe ++ [ ChildByIndex 1 ]) output)
+
+        Type.Unit a ->
+            Type.Unit (mapAttribute a)
+
+
+{-| Map type attribute
+-}
+mapTypeAttributeWithNodePath : (NodePath -> attr -> attr2) -> Type attr -> Type attr2
```
Comment (AttilaMihaly) [inline src/Morphir/IR/NodeId.elm:318]: Needs more explanation.

File: src/Morphir/IR/NodeId.elm [hunk 9]
```diff
@@ -1,48 +1,795 @@
...
+                    getValueAttributeByPath nodePath v
+
+                Nothing ->
+                    Err <| InvalidNodeID (nodeIdToString nodeId)
+
+
+{-| Get attribbute from a module, type or value by nodeID
+Return erro if the NodeID is invalid
+-}
+mapTypeAttributeWithNodePathRec : (NodePath -> attr -> attr2) -> NodePath -> Type attr -> Type attr2
```
Comment (AttilaMihaly) [inline src/Morphir/IR/NodeId.elm:268]: Punctuation is missing. I don't see how this function would return an error.

File: src/Morphir/IR/NodeId.elm [hunk 10]
```diff
@@ -1,48 +1,795 @@
...
+                |> List.Extra.getAt n
+                |> Maybe.map (recursiveFunction xs (List.append [ current ] pathSoFar))
+                |> Maybe.withDefault (returnInvalidPathError pathSoFar)
+
+        _ ->
+            returnInvalidPathError pathSoFar
+
+
+{-| Get attribbute from a moudle, type or value by nodeID
+Return erro if the NodeID is invalid
```
Comment (AttilaMihaly) [inline src/Morphir/IR/NodeId.elm:216]: Punctuation is missing.

**Review by AttilaMihaly — CHANGES_REQUESTED** [review]

File: src/Morphir/CustomAttribute/CustomAttribute.elm
Comment (AttilaMihaly) [inline src/Morphir/CustomAttribute/CustomAttribute.elm:1]: Is this still necessary? I thought we moved everything to Decorations.

**Review by AttilaMihaly — CHANGES_REQUESTED** [review]

File: src/Morphir/CustomAttribute/CustomAttribute.elm
Comment (bekand) [reply src/Morphir/CustomAttribute/CustomAttribute.elm:1]: I think you are right, that file was unused. I removed it.

**Review by AttilaMihaly — CHANGES_REQUESTED** [review]
There's one cosmetic change left.

**Pull request merged**